### PR TITLE
Fix invalid triggers of the uncaughtException handler

### PR DESCRIPTION
**Pull request by mridgway**
## Pre-Submission Checklist

- [x] Opened an issue discussing these changes before opening the PR
- [x] Ran the linter and tests via `make prepush`
- [x] Included comprehensive and convincing tests for changes

## Issues

Closes:

* Issue #1709 

# Changes

The previous (prior to 7.x) functionality for uncaught exceptions only triggered the `uncaughtException` handler if it was truly an uncaught thrown error. Right now in Restify 7, if you turn on `handleUncaughtExceptions` and add a listener for the event, all errors from `next(err)` and internal errors like `MethodNotAllowed` will trigger the `uncaughtException` handler.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -89,7 +89,8 @@ function createServer(options) {
             e
         ) {
             if (
-                this.listeners('uncaughtException').length > 1 ||
+                (opts.handleUncaughtExceptions &&
```
Comment (DonutEspresso) [inline lib/index.js]: Isn't this already gated (`opts.handleUncaughtExceptions`) on L84?

File: test/server.test.js [hunk 2]
```diff
@@ -674,6 +674,42 @@ test('GH-77 uncaughtException (default behavior)', function(t) {
     });
 });
 
+// eslint-disable-next-line
+test('handleUncaughtExceptions should not call handler for internal errors', function(t) {
+    SERVER.get('/', function(req, res, next) {
```
Comment (DonutEspresso) [inline test/server.test.js]: I'm a little bit confused about this test setup. I think we trying to verify restify/router errors (e.g., 404) don't trigger uncaught exceptions? If yes, should we just remove the get route and have the client do a get against a non-existent route?

File: test/server.test.js [hunk 3]
```diff
@@ -674,6 +674,42 @@ test('GH-77 uncaughtException (default behavior)', function(t) {
...
+
+    CLIENT.head('/', function(err, _, res) {
+        t.ok(err);
+        t.equal(res.statusCode, 405);
+        t.end();
+    });
+});
+
+// eslint-disable-next-line
+test('handleUncaughtExceptions should not call handler for next(new Error())', function(t) {
```
Comment (DonutEspresso) [inline test/server.test.js]: :+1:

File: test/server.test.js (hunk 2 above)
Comment (mridgway) [reply test/server.test.js]: There seems to be a behavior difference if you have no routes (which 404s) and having only GET routes and trying HEAD. I agree that this makes the test confusing, but I think if the behavior difference should be fixed, it should be fixed separately.

File: lib/index.js (hunk 1 above)
Comment (mridgway) [reply lib/index.js]: Yeah, it is indeed. I'll remove this check.

File: test/server.test.js (hunk 2 above)
Comment (DonutEspresso) [reply test/server.test.js]: Gotcha - thanks for context! I expect this is related to CORS use cases - do you mind adding all of this as a comment to the test? Otherwise, it's a bit confusing to register the empty `/` that we actually don't expect to be run or triggered. Maybe we could do an `assert.fail('should not run')` or something inside the GET handler for clarity as well?

Comment (mridgway) [reply test/server.test.js]: Makes sense. I added a comment and the assertion that you mentioned.

File: lib/server.js [hunk 4]
```diff
@@ -816,7 +816,7 @@ Server.prototype._onRequest = function _onRequest(req, res) {
         handlerDomain.add(req);
         handlerDomain.add(res);
         handlerDomain.on('error', function onError(err) {
-            self._onHandlerError(err, req, res);
+            self._onHandlerError(err, req, res, true);
```
Comment (misterdjules) [inline lib/server.js]: I think this would still make errors _emitted_ on objects bound to the domain (in this case `req` and `res`) trigger an `uncaughtException` event. It seems that would also be a problem?

If we consider that is indeed not what the behavior of `handleUncaughtExceptions` should be, we should set the fourth parameter of `_onHandlerError` to `true` only if `err.domainThrown === true`.

@mridgway @DonutEspresso Thoughts?

Comment (mridgway) [reply lib/server.js]: I believe that behavior is consistent with Restify versions <7. I'm not sure what the original intention was though.

I agree that it would make more sense to filter to only uncaught exceptions.

Comment (misterdjules) [reply lib/server.js]: > I believe that behavior is consistent with Restify versions <7.

I just verified that looking at the 6.x branch of this repository and this is correct. Thus I'm fine with keeping that behavior, at least for now.
